sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -468,13 +458,25 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
       throw new ParseException("There must be at least one WHEN clause in a MERGE statement", ctx)
     }
     // children being empty means that the condition is not set
-    if (matchedActions.length == 2 && matchedActions.head.children.isEmpty) {
-      throw new ParseException("When there are 2 MATCHED clauses in a MERGE statement, " +
-        "the first MATCHED clause must have a condition", ctx)
-    }
-    if (matchedActions.groupBy(_.getClass).mapValues(_.size).exists(_._2 > 1)) {
+    val matchedActionSize = matchedActions.length
+    if (matchedActionSize >= 2 && !matchedActions.init.forall(_.condition.nonEmpty)) {
```

Comment:
we can still write `.children.isEmpty`, then we don't need to change `v2Commands.scala`

Reply:
I don't think so, because the `children` of `InsertAction` and `UpdateAction` actually include `condition` and `assignments`. There may be cases where there're `assignments` and `condition` being ignored but `children` is nonEmpty.

Reply:
then this is an existing bug?

Reply:
Yes, it was a bug.

Reply:
can you send a new PR to fix this bug? we need to fix it in 3.0 branch as well.

Reply:
Submitted a pr at https://github.com/apache/spark/pull/28943.